Implement drag and drop for maintable

Siedlerchr opened the pull request:
Drag And Drop Functionality
- [X]  MainTable to Group
~~- [ ] MainTable to MainTable (For reodering?Useful?)~~ Not useful
- -  [x] Expand group node and children on drag over 

- [x]  Files to Maintable 
- -  [x] Add File to Entry
- -  [x] Create new Entry from PDF/XMP
- -  [ ] Show Merge dialog if entry already exists


- [x] Files to EntryEditor (mostly PDF)
- - [x] fix folder with filename is created #4087
- [x] Files to Preview
- [x] `.bib` files to maintable -> Import
- [x] `.bib` files to tabbed pane -> open 


Fixes #4086 
Fixes #3923 
Fixes #3258 
Fixes #4087 

Current behaviour Windows: Shift or no modifer on maintable (move)
 - Create new entry from XMP content (if pdf)
    -    XMP content found: Show dialog with content and option to import entry or to create file
<kbd>ALT</kbd> on maintable: move file to file dir and link (only windows)
<kbd>CTRL</kbd> on maintable: copy file to file dir (if not exists)  and link
Bib-File dragged to maintable: Import contents

Xubuntu: No modifier (Copy)
  - Create new entry from XMP content (if pdf)
    -    XMP content found: Show dialog with content and option to import entry or to create file
<kbd>Shift </kbd> on maintable :  move to file dir 

Preview Panel works the same as above except it doesn't handle bib files import,

BibFile dragged below tab pane/table header -> Open bib libray 

General Tab Files -> Link file 


----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

tobiasdiez commented:
Cool that you have a look at the drag'n'drop functionality. It's indeed one of the major things missing in the new maintable.
I think, there are generally three different scenarios:
- Source outside of JabRef, Target maintable: This may be a pdf file (-> import as new entry), some string (-> try to parse/import) or a bib file (-> import and if target is not the maintable but the tab header -> open as new library). 
- Target outside of JabRef, Source maintable: For this a string representation is needed (since no external application can handle BibEntries) but I'm not sure what we should serialize. The bibtex code seems logical but also the preview format (e.g. drag'n'drop to email program) or `\cite{key}̀  (target latex editor) makes sense. Maybe add a user preference?
- Target and source are JabRef nodes: Then it is in general desirable to not serialize the entry but pass it by reference. For example, drag an entry from the maintable to the group pane; one would need the original entry to modify the `groups` field. Not sure if we need drag'n'drop from maintable to maintable (what's the use for this?). For the implementation of these local operations by reference, I would use the following code as a starting point: https://community.oracle.com/message/10909582#10909582 (slightly modified to use Optionals instead of nulls).

Siedlerchr commented:
I implemented drag and drop a while ago for the Linked Files editor, so I know now to add serialization. Drag and drop from and to maintable could be used for reordering entries. But not sure if that makes sense. I just implemented it to test if it works in general

lenhard commented:
I agree with everything that @tobiasdiez writes. But just for the record: Turning `BibEntry` serializable is trivial. It's just a marker interface and there is no functionality to add. Just a few attributes should be marked as `transient` and that's it. Nevertheless, I think it's better to serialize the BibTeX String representation.

Siedlerchr commented:
@lenhard  In theory es, but in practice I now stumbled across the problem (again) that those javafx ObservableMaps/List/Sets etc can't be serialized. So I tried adding custom readObject/writeObject methods, but still can't serialize it, so I would prefer sticking with the parser/serialization.

Ideally we would have a method which is the exact opposite of the serializeAll() - e.g. deserialzeAll() without having to use the parser overhead. Or maybe a simple toString() - reverse 

And I come to the conclusion that this is not the best idea and we should use some trans

tobiasdiez commented:
Yes, the javafx properties are never serializable because you cannot serialize listeners (they same applies to the event bus). But as I argued above, the best is to never serialize any Java object if drag'n'drop happens between nodes in JabRef itself. Just store a reference to the objects that are dragged.

Siedlerchr commented:
Okay, I will try with that LocalClipboard/DragBoard Implementation

Siedlerchr commented on src/main/java/org/jabref/model/database/BibDatabase.java:
```diff
@@ -116,7 +116,7 @@ public boolean containsEntryWithId(String id) {
 
     public ObservableList<BibEntry> getEntries() {
-        return FXCollections.unmodifiableObservableList(entries);
+        return entries;
```
Is there a reason why the entries are all wrapped in an Unmodiefable collection?

tobiasdiez replied:
Yes, so that nobody can add entries to the database without calling `insertEntry` which also does some bookkeeping (id list, tex free version, ...). I think the general principle is that objects should never expose their inner data.

Siedlerchr commented on src/main/java/org/jabref/gui/maintable/MainTable.java:
```diff
@@ -279,62 +278,75 @@ public void paste() {
+            int draggedIndex = (Integer) db.getContent(SERIALIZED_MIME_TYPE);
 
-        ObservableList<BibEntryTableViewModel> items = this.itemsProperty().get();
+            BibEntry entry = this.getItems().get(draggedIndex).getEntry();
```
I followed the guide in https://stackoverflow.com/a/28606524/3450689 
@tobiasdiez  How do I implement the insert/delete at best? The problem is that all those ViewModel stuff is wrapped in many Unmodiefiable/ReadOnly Collections which don't support remove/insert 

And calling database.removeEntry/insertEntry everytime I just do drag and drop on the maintable seems to be a bit overhead (as all kind of events are triggering)
Or do we just don't allow reordering the MainTable?

tobiasdiez replied:
I think reordering of entries is a delicate business: what does it mean to reorder entries if I sort the table by author? The order of the entries in the bib file is determined independently anyway. For these reasons, I would not try to implement reordering of entries.

koppor commented:
@Siedlerchr I added "`.bib` files to maintable. - Current behavior: dropping to tab bar works, but that just reorders the tab bar, but does not open the `.bib` file.".

Siedlerchr commented:
I'll take a look. currently I only check for files in maintable. So when I drop a bib file to the maintabe it should openn in a new tab?

koppor commented:
Yes :)

tobiasdiez commented:
If you drop onto the maintable it should import the content of the bib file into the current database. If the drop target is the tab bar, it should open the bib file as a new tab.

Siedlerchr commented:
@koppor  You might want to check out the updated description in the issue and test the current version.  In Windows, the default drag and drop  is move which also applies when using the SHIFT key. So, both result in the same event

tobiasdiez commented:
It would be nice to give visual feedback about the drop action: https://docs.oracle.com/javase/8/javafx/events-tutorial/drag-drop.htm#BABFACBF (instead of changing the color in code, add/remove a pseudo-class and style it using css).

AEgit commented:
The following two issues might be relevant to this as well:

https://github.com/JabRef/jabref/issues/3258
https://github.com/JabRef/jabref/issues/2869

Siedlerchr commented:
@tobiasdiez  The main feedback I wanted was about the functionality of the modifiers and the options. 
Regarding the log statements, of course they will be deleted in final commit. 

Regarding the Local Dragboard,  the current implementation is essentially a static guaranteed thread safe singleton 
I see no advantage over initializing this as a static field which would suffer from potential thread safety on initialization, especially in the drag and drop functionality.

tobiasdiez commented:
As we try to get ride of all singletons and static fields in `Globals`, both solutions are not perfect. But using a static field and dependency injection as much as possible should be easier to refactor than a singleton that is used everywhere directly. Threading should not be an issue as every drag & drop event should be thrown in the JavaFX thread anyway.

Siedlerchr commented:
I tested and updated the results with my observance under Xubuntu. No modifier under Ubuntu equls TransferMode.Copy
Shift on Xubuntu:  Transfermode.Move

Windows: No modifier and or shift: Transfermode.Move
CTRL Modifier : Transfermode Copy

Siedlerchr commented:
@tobiasdiez  I still don't really get how I the NewDroppedFileHandler could be changed, because they all act on dropped files. Either they import the entries or if no entries are found the file is added. So I need both methods, or if I extract the latter ones to a new class, I still would need to call the methods for moving files. Maybe you have a better idea

tobiasdiez commented:
My proposal was to split it as follows:
- `EntryFileLinker`: provides a bunch of methods that allow to move/copy/link files to a given BibEntry (which was passed as a constructor argument)
- `DatabaseFileLinker`: allows to add files to a given database (passed as constructor argument) and generate BibEntries out of the content. You are of course allowed to reuse `EntryFileLinker` if this is convenient.
- GUI stuff for the dialogs

(the names for the classes are not very good). The first two classes should reside in logic. But maybe I miss something why such a splitting is not possible.

tobiasdiez commented on src/main/java/org/jabref/model/entry/BibEntry.java:
```diff
@@ -57,14 +57,15 @@
      */
     private final Map<String, String> latexFreeFields = new ConcurrentHashMap<>();
-    private final EventBus eventBus = new EventBus();
+    private final transient EventBus eventBus = new EventBus();
```
no longer needed right?

tobiasdiez commented on src/main/java/org/jabref/model/entry/SharedBibEntryData.java:
```diff
@@ -1,9 +1,11 @@
  *  Stores all informations needed to manage entries on a shared (SQL) database.
  */
-public class SharedBibEntryData {
+public class SharedBibEntryData implements Serializable {
```
̀Serializable should also no longer be needed.

Siedlerchr commented on src/main/java/org/jabref/gui/util/CustomLocalDragboard.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class CustomLocalDragboard {
+
+    private final Map<Class<?>, Object> contents = new HashMap<>();
```
With this version we have a complete generic type safe local dragboard. Implementation for groups and bib entries would be the essentially the same code so I decided to use a generic type safe approach.In addition provides possibilities for all other kind of types

tobiasdiez replied:
There speaks nothing against leaving the generic Map here, but personally I find `putValue(DragAndDropDataFormats.BIBENTRY_LIST_CLASS, entries)` harder to read than `putEntries(entries)`. Thus, my suggestion was to add overloads for the common use cases (similar to how the javafx dragboard has get/has/putFiles methods).

Siedlerchr commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -215,6 +219,30 @@ private void init() {
         initKeyBindings();
 
+        tabbedPane.setOnDragOver(event -> {
+            if (event.getDragboard().hasFiles() && (event.getSource() instanceof DndTabPane)) {
```
No longer needed.

Siedlerchr commented on src/main/java/org/jabref/gui/PreviewPanel.java:
```diff
@@ -89,17 +105,57 @@ public PreviewPanel(BasePanel panel, BibDatabaseContext databaseContext, KeyBind
+                if (event.getTransferMode() == TransferMode.MOVE) {
+
+                    System.out.println("Mode MOVE"); //shift on win or no modifier
+                    fileHandler.addNewEntryFromXMPorPDFContent(entry, files);
```
I thought the default drag and drop operation should be importing the content? This was discussed in #121

tobiasdiez replied:
According to: https://github.com/JabRef/jabref/issues/121#issuecomment-368215205
- Drop on preview/entry editor: link file to target entry
- Drop on preview/entry editor with Alt held: link file to target entry
- Drop on preview/entry editor with Shift held: rename to filename pattern, move the file to file directory and link file to target entry

Siedlerchr replied:
The problem is that 1 and 3 are the same on Windows.  so I am going to stick with option 3.

tobiasdiez commented:
Thanks for the quick follow-up. Merging, since I'm somewhat trigger happy right now ;-)
